python/lsst/daf/butler/registries/sqlRegistry.py

```diff
@@ -278,88 +318,54 @@ def query(self, sql, **params):
 
     @transactional
     def registerDatasetType(self, datasetType):
-        """
-        Add a new `DatasetType` to the SqlRegistry.
-
-        It is not an error to register the same `DatasetType` twice.
-
-        Parameters
-        ----------
-        datasetType : `DatasetType`
-            The `DatasetType` to be added.
-
-        Raises
-        ------
-        ValueError
-            DatasetType is not valid for this registry or is already registered
-            but not identical.
-
-        Returns
-        -------
-        inserted : `bool`
-            `True` if ``datasetType`` was inserted, `False` if an identical
-            existing `DatsetType` was found.
-        """
-        if not self._isValidDatasetType(datasetType):
-            raise ValueError("DatasetType is not valid for this registry")
-        # If a DatasetType is already registered it must be identical
-        try:
-            # A DatasetType entry with this name may exist, get it first.
-            # Note that we can't just look in the cache, because it may not be there yet.
-            existingDatasetType = self.getDatasetType(datasetType.name)
-        except KeyError:
-            # No registered DatasetType with this name exists, move on to inserting it
-            pass
-        else:
-            # A DatasetType with this name exists, check if is equal
-            if datasetType == existingDatasetType:
-                return False
+        # Docstring inherited from Registry.getDatasetType.
+        datasetType.normalize(universe=self.dimensions)
+        # If the DatasetType is already in the cache, we assume it's already in
+        # the DB (note that we don't actually provide a way to remove them from
+        # the DB).
+        existingDatasetType = self._datasetTypes.get(datasetType.name, None)
+        # If it's not in the cache, try to insert it.
+        if existingDatasetType is None:
+            try:
+                self._connection.execute(
+                    self._schema.tables["DatasetType"].insert().values(
```

Comment:
There is already a `begin` from the `transactional` decorator.

Reply:
Just a note - in "true" databases which use different transaction isolation modes this conflict could happen during commit phase, not necessarily at insert.

Reply:
Well, even for upsert I think concurrency conflicts may be detected only on commit, again depending on transaction isolation and locking model used by particular database.

Reply:
Yes; my hope was that with an UPSERT we can tell the database directly how to resolve the conflict based on which of two constraints was violated, and then we don't care when that conflict resolution happens.  I'm not certain that's possible in general, but I think we've demonstrated that it can work with both SQLite and Oracle, even if (in the former case, at least) we're not sure if it's documented, guaranteed behavior or just luck that the implementation happens to do what we want.

Reply:
Oracle does not wait for commits to detect and resolve deadlocks. I don't really know how to see the SQL that is causing this without being able to run it in Oracle but, in general, you can avoid deadlocks if your transactions and resource locking all happen in the same direction (you all probably already know that but just in case).  

Still trying to figure out if @andy-slac considers Oracle a "true" database. :)

Reply:
Oracle is definitely much truer database than SQLite 😉